ui/pages/confirmations/hooks/alerts/useDomainMismatchAlert.ts

```diff
@@ -0,0 +1,46 @@
+import { useMemo } from 'react';
+import {
+  isValidSIWEOrigin,
+  WrappedSIWERequest
+} from '@metamask/controller-utils';
+
+import { Alert } from '../../../../ducks/confirm-alerts/confirm-alerts';
+import { Severity } from '../../../../helpers/constants/design-system';
+import { useI18nContext } from '../../../../hooks/useI18nContext';
+import { SignatureRequestType } from '../../types/confirm';
+import { isSIWESignatureRequest } from '../../utils';
+import useCurrentConfirmation from '../useCurrentConfirmation';
+
+const useDomainMismatchAlerts = (): Alert[] => {
+  const { currentConfirmation } = useCurrentConfirmation();
+  const t = useI18nContext();
+
+  const alerts = useMemo<Alert[]>(() => {
+    const isSIWE = isSIWESignatureRequest(
+      currentConfirmation as SignatureRequestType
+    );
+    if (!isSIWE || !currentConfirmation?.msgParams) {
+      return [];
+    }
+
+    const { msgParams } = currentConfirmation;
+    const isSIWEDomainValid = isValidSIWEOrigin(msgParams as WrappedSIWERequest);
+    if (isSIWEDomainValid) {
+      return [];
+    }
+
+    const alert = {
```

Comment:
Since none of the logic up to this point has any meaningful performance cost, could we move of all this out of the `useMemo`?

This has a few benefits, firstly the readability is improved since we can nest only the logic that needs to be cached, plus we no longer need to re-generate the alert array every time the confirmation ID or params change, but instead we can do so specifically when the alert is no longer needed.

For example, we could generate a `isDomainMismatch` boolean and pass that as the only dependency.

Reply:
Thanks for the explanation with the suggestion :)  

good point we wouldn't need to regenerate the array on the other prop changes. Updated [78363ec](https://github.com/MetaMask/metamask-extension/pull/25281/commits/78363ec074bf2ec07d9aa2045ebd6df9a0a76a52)

I don't quite follow the last sentence. as in pass a boolean as a param which could control the change?

Reply:
Sorry for the confusion! We still need a `useMemo` to hold onto the array reference to avoid re-renders, but the only dependency we need is a boolean such as `isDomainMismatch` which controls whether we return an empty array or the alert.

This is the pattern used throughout the transaction alerts PR for reference.

Reply:
okay actually this makes more sense now

[8fcaec9](https://github.com/MetaMask/metamask-extension/pull/25281/commits/8fcaec92f7a393b70da245fb7aa8b0dbda719fb3)